### PR TITLE
Add compile-time integer overflow checking for bounds widening

### PR DESCRIPTION
**Pull request by mgrang**
In order to make an expression e uniform we split the expr into a "base" and an
"offset". The base contains all DeclRefExprs of e and the Offset contains all
IntegerLiterals of e. In order to do the split, we need to re-associate the
integers.

For example: We convert an expression like (p + i) + j into p + (i + j), where
p is DeclRefExpr and i and j are integers. In this case, (i + j) may
potentially overflow. We implement compile time checking to check for integer
overflow.

### CODE REVIEW

**Comment by mgrang** [comment]
This is a follow up to #765

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 1]
```diff
@@ -304,8 +304,12 @@ void BoundsAnalysis::FillGenSetAndGetBoundsVars(const Expr *E,
     // relative to the declared upper bound expression. This offset is used in
     // the widening computation in ComputeOutSets.
 
-    // TODO: Check to see if that difference overflows/underflows.
-    EB->Gen[SuccEB->Block][V] = (DerefOffset - UpperOffset).getLimitedValue();
+    // Check if the difference overflows.
+    bool Overflow;
+    UpperOffset = DerefOffset.ssub_ov(UpperOffset, Overflow);
+    if (Overflow)
```
Comment (mgrang) [inline clang/lib/Sema/BoundsAnalysis.cpp]: Should we simply skip widening on overflow or should we assert?

File: clang/lib/Sema/BoundsAnalysis.cpp [hunk 2]
```diff
@@ -398,9 +402,14 @@ ExprIntPairTy BoundsAnalysis::SplitIntoBaseOffset(const Expr *E) {
...
     // (p + j) + i ==> return (p, j + i)
 
-    // TODO: Since we are reassociating integers here, check if the value
-    // overflows/underflows.
-    return std::make_pair(BinOpLHS, IntVal + BinOpRHS);
+    // Since we are reassociating integers here, check if the value
+    // overflows.
+    bool Overflow;
+    IntVal = IntVal.sadd_ov(BinOpRHS, Overflow);
+    if (Overflow)
```
Comment (mgrang) [inline clang/lib/Sema/BoundsAnalysis.cpp]: Should we simply return `(nullptr, zero)` on overflow or should we assert?

File: clang/test/CheckedC/inferred-bounds/widened-bounds.c [hunk 3]
```diff
@@ -464,3 +466,72 @@ void f19() {
...
+  if (*(r + INT_MAX))                               // expected-error {{out-of-bounds memory access}}
+  {}
+
+// CHECK:  [B8]
+// CHECK:    1: _Nt_array_ptr<char> r : count((-2147483647 - 1)) = "";
+// CHECK:    2: *(r + 2147483647)
+// CHECK:  [B7]
+// CHECK-NOT: upper_bound(r)
+
+  // Declared bounds and deref offset (INT_MAX + 1). Integer overflow. No widenening.
```
Comment (dtarditi) [inline clang/test/CheckedC/inferred-bounds/widened-bounds.c]: There a typo in the comment and a few other comments: `widenening` -> `widening`.

**Comment by mgrang** [comment]
Hard coding the value of INT_MAX in tests may result in the tests failing on 64-bit architectures. I am working on fixing this.

File: clang/test/CheckedC/inferred-bounds/widened-bounds.c [hunk 4]
```diff
@@ -464,3 +466,66 @@ void f19() {
...
+  {}
+
+// CHECK:  [B10]
+// CHECK:    2: *(q + {{.*}})
+// CHECK:  [B9]
+// CHECK: upper_bound(q) = 1
+
+  // Declared bounds (INT_MIN) and deref offset (INT_MAX - 1). No sequential deref tests. No widening.
+  _Nt_array_ptr<char> r : count(INT_MIN) = "";
+  if (*(r + INT_MAX - 1))                               // expected-error {{out-of-bounds memory access}}
```
Comment (mgrang) [inline clang/test/CheckedC/inferred-bounds/widened-bounds.c]: Windows X86 Debug build fails to display the error `out-of-bounds memory access`. This seems to happen only at `*(p + INT_MAX)`. So for now, I have changed the dereference to `*(p + INT_MAX - 1)` to make this test pass. I have filed isue https://github.com/microsoft/checkedc-clang/issues/780. This needs to be investigated. I will come back and fix it later.

**Comment by mgrang** [comment]
@dtarditi The updated PR is ready for review. ADO passed.

File: clang/test/CheckedC/inferred-bounds/widened-bounds.c (hunk 4 above)
Comment (dtarditi) [reply clang/test/CheckedC/inferred-bounds/widened-bounds.c]: Could you add a TODO before line 494 that includes the GitHub issue?

**Comment by mgrang** [comment]
TODO added. Merging PR.

**Pull request merged**